libp2p/transports/tcptransport.nim

```diff
@@ -18,52 +19,103 @@ import transport,
 logScope:
   topic = "TcpTransport"
 
-type TcpTransport* = ref object of Transport
-  server*: StreamServer
+const
+  TcpTransportTrackerName* = "libp2p.tcptransport"
+
+type
+  TcpTransport* = ref object of Transport
+    server*: StreamServer
+    cleanups*: seq[Future[void]]
+    handlers*: seq[Future[void]]
+
+  TcpTransportTracker* = ref object of TrackerBase
+    opened*: uint64
+    closed*: uint64
+
+proc setupTcpTransportTracker(): TcpTransportTracker {.gcsafe.}
+
+proc getTcpTransportTracker(): TcpTransportTracker {.gcsafe.} =
+  result = cast[TcpTransportTracker](getTracker(TcpTransportTrackerName))
+  if isNil(result):
+    result = setupTcpTransportTracker()
+
+proc dumpTracking(): string {.gcsafe.} =
+  var tracker = getTcpTransportTracker()
+  result = "Opened transports: " & $tracker.opened & "\n" &
+           "Closed transports: " & $tracker.closed
+
+proc leakTransport(): bool {.gcsafe.} =
+  var tracker = getTcpTransportTracker()
+  result = (tracker.opened != tracker.closed)
+
+proc setupTcpTransportTracker(): TcpTransportTracker =
+  result = new TcpTransportTracker
+  result.opened = 0
+  result.closed = 0
+  result.dump = dumpTracking
+  result.isLeaked = leakTransport
+  addTracker(TcpTransportTrackerName, result)
 
 proc cleanup(t: Transport, conn: Connection) {.async.} =
   await conn.closeEvent.wait()
+  trace "connection cleanup event wait ended"
   t.connections.keepItIf(it != conn)
 
-proc connHandler*(t: Transport,
+proc connHandler*(t: TcpTransport,
                   server: StreamServer,
                   client: StreamTransport,
-                  initiator: bool = false):
-                  Future[Connection] {.async, gcsafe.} =
+                  initiator: bool): Connection =
   trace "handling connection for", address = $client.remoteAddress
   let conn: Connection = newConnection(newChronosStream(server, client))
   conn.observedAddrs = MultiAddress.init(client.remoteAddress)
   if not initiator:
     if not isNil(t.handler):
-      asyncCheck t.handler(conn)
+      t.handlers &= t.handler(conn)
 
     t.connections.add(conn)
-    asyncCheck t.cleanup(conn)
+    t.cleanups &= t.cleanup(conn)
 
   result = conn
 
 proc connCb(server: StreamServer,
             client: StreamTransport) {.async, gcsafe.} =
   trace "incomming connection for", address = $client.remoteAddress
-  let t: Transport = cast[Transport](server.udata)
-  asyncCheck t.connHandler(server, client)
+  let t = cast[TcpTransport](server.udata)
+  # we don't need result connection in this case
+  # as it's added inside connHandler
+  discard t.connHandler(server, client, false)
```

Comment:
It's better to leave `asyncCheck` here then, it's much safe than simply discarding.

Reply:
`connHandler` is not `async` so there is no need
the resulting connection is stored in `connHandler` in `t.connections` so we don't need it either.

The caller of `connCb` will activate all the futures quequed in `connHandler` btw.

Reply:
Ah, you right, then we're good :)